### DISCUSSION TITLE
Invoicing and marking time as "paid"

### DISCUSSION THREAD

**Discussion by maxiride** [jotaen/klog#65] [Help / Question]
I was browsing for new tools for time tracking and klog popped out.
In the scenario of using klog to track working time for freelance work, is there a way to mark timespans as billed?

The use of tags seems a good starting point for this use case either applying them to the overall record or single timespans, however, it seems to me that the --tag cli command is only to include tags.

1. Is there a way to query all records without a tag? eg. print me all the records without #billed
2. Is there a way to query records that have\have not certain tags but not others? eg. print me all the records #client1 with\without #billed
3. Is there a way to bulk apply tags to records? eh. apply the #billed tag to all the records that have the tags #client1 #projectA

**Comment by vladdeSV** [comment]
To answer your first question: You can use `--since 2021-01-01 --until 2021-02-01` or `--after 2021-01-01 --before 2021-02-01`

Eg. 
```sh
klog total ./time.klg --since 2021-03-01 --before 2021-04-01
```

**Reply by maxiride** [reply]
I'm sorry, but those arguments apply to dates, how would they affect tags?

> 1. Is there a way to query all records without a tag? eg. print me all the records without #billed

I meant to query klog for all records without a tag regardless of the record's date.

**Reply by vladdeSV** [reply]
Excuse me, I misread the question. Unfortunately I do not have a proper answer to your question.

**Comment by jotaen** [comment]
> Is there a way to query all records without a tag? eg. print me all the records without #billed

No, currently the filters are quite simple, so you can’t do things along the lines of `OR` or `NOT`. I am open to making the filters more powerful, but so far I couldn’t think of an easy and “CLI-native” way to do that, and I don’t want to build a query language 😁

> Is there a way to query records that have\have not certain tags but not others? eg. print me all the records #client1 with\without #billed

No, see above.

> Is there a way to bulk apply tags to records? eh. apply the #billed tag to all the records that have the tags #client1 #projectA

No, the file manipulations via the CLI are intentionally simple. I would say that once you cannot conveniently edit your files through a text editor anymore, you probably are better off with another time tracker that supports more complex use-cases. The klog CLI is mostly a helper / convenience for interacting with your plain text data.

I don’t know about your specific setup, but maybe it’s viable for you to split up your files, e.g. one file per invoice-period and client. For evaluation you can still mix them together, e.g. `klog total client1/2021-04.klg client2/2021-04.klog` or `klog total */*-04.klg` (glob) or the like.

**Reply by maxiride** [reply]
> No, currently the filters are quite simple, so you can’t do things along the lines of OR or NOT. I am open to making the filters more powerful, but so far I couldn’t think of an easy and “CLI-native” way to do that, and I don’t want to build a query language 😁


Would it be simple, and most importantly effective, to simply add an argument --exclude-tag=comma,separated,taglist?

>I don’t know about your specific setup, but maybe it’s viable for you to split up your files, e.g. one file per invoice-period and client. For evaluation you can still mix them together, e.g. klog total client1/2021-04.klg client2/2021-04.klog or klog total */*-04.klg (glob) or the like.

Indeed, nice suggestion.

**Reply by jotaen** [reply]
> Would it be simple, and most importantly effective, to simply add an argument --exclude-tag=comma,separated,taglist?

It would do the trick for the tags, but one could make a point that this should also be supported for the other filters, e.g. “give me all records from 2018, but not from this one week, plus everything that is tagged with #xyz regardless of the date”. This pattern also couldn’t be applied to `OR`-ing tags, e.g. “give me all records tagged with #client1 or #client2”

I guess I need to think about it some more…
